### ISSUE TITLE
Windows progress

### ISSUE THREAD

**Issue by zetashift** [greenfork/nimraylib_now#1]
First off, thank you for your wonderful work!
I've been tinkering on Windows with this, and anything that requires `raylib.dll` works, atleast the samples I tried. I'll be playing with this more!

1. Download the Windows release from: https://github.com/raysan5/raylib/releases, but don't choose the installer, choose the other one, like for the 3.5.0 release you download: `raylib-3.5.0_win64_mingw-w64.zip` if you are on a 64bits system or `raylib-3.5.0_win32_mingw-w64.zip` if you are on a 32bits system. 
2. In the downloaded archive there is a `/lib` folder in there you'll find `raylib.dll`,  put this dll in the same folder as your resulting `.exe` file.

So say we want to run this https://github.com/greenfork/nimraylib_now/blob/master/examples/shapes/bouncing_ball.nim example. `cd` yourself into the folder, run `nim c -r .\bouncing_ball.nim` and make sure the dll is in the same folder, and you'll find yourself seeing this:

![screenshot_20210110_143702](https://user-images.githubusercontent.com/1857826/104124531-aaec4780-5351-11eb-98d9-8b248b3179a7.png)

### Testing doesn't work

Running `nimble testExamples` gives the following error
```
PS P:\dev\gamedev\nimraylib_now> nimble testExamples
  Executing task testExamples in P:\dev\gamedev\nimraylib_now\nimraylib_now.nimble
no tests were found for pattern: 'tests/examples/**/*.nim'
```
Maybe this is has to do with how paths are handled differently on Windows and Linux?

### Raygui
I downloaded raygui.dll from your release page, put it next to raylib.dll but the example still fails with:
` could not load: raygui.dll `
Running `buildRaygui` gives me the following error:
```
  Executing task buildRaygui in P:\dev\gamedev\nimraylib_now\nimraylib_now.nimble
stack trace: (most recent call last)
C:\Users\Rishi\AppData\Local\Temp\nimblecache-0\nimscriptapi_2903036856.nim(187, 16)
P:\dev\gamedev\nimraylib_now\nimraylib_now.nimble(26, 8) buildRayguiTask
C:\Users\Rishi\.choosenim\toolchains\nim-1.4.2\lib\system\nimscript.nim(186, 19) exec
C:\Users\Rishi\.choosenim\toolchains\nim-1.4.2\lib\system\nimscript.nim(186, 19) Error: unhandled exception: The system cannot find the file specified.
 [OSError]
     Error: Exception raised during nimble script execution
```
This might have to do with me not having raygui.o anywhere. I'll try to troubleshoot this tonight or tomorrow.

**Comment by greenfork** [comment]
Thanks very much for the docs, this will definitely go to readme!

I fixed the issue with `/` so that it should be `\` on Windows, this should solve testing.

With raygui it is a bit harder, the exact commands in [nimble file](https://github.com/greenfork/nimraylib_now/blob/master/nimraylib_now.nimble#L33-L37) are:

```nim
  let sourceFile = "raygui"/"src"/"raygui.h"
  exec "cc -x c -fPIC -c -o raygui.o -DRAYGUI_IMPLEMENTATION -DRAYGUI_SUPPORT_ICONS " & sourceFile
  exec "cc -shared -o libraygui.so -lraylib raygui.o"
```

On windows at least `-o libraygui.so` should be `-o raygui.dll` (as per conventions) and God knows what else. I searched for "how to compile a dynamic library for Windows" and got:
- https://docs.microsoft.com/en-us/cpp/build/walkthrough-creating-and-using-a-dynamic-link-library-cpp?view=msvc-160
- https://www.codeproject.com/Articles/84461/MinGW-Static-and-Dynamic-Libraries

The commands should be similar though. And raygui.h requires raylib.h to be in path, you can just download it from raylib if it's not loaded by default https://github.com/raysan5/raylib/blob/master/src/raylib.h

**Comment by greenfork** [comment]
Could you test raygui usage with `-d:nimDebugDlOpen`, it should give more info on what's actually wrong with this dll. I personally took it from here: https://github.com/Guevara-chan/Raylib-Forever/releases
```
nim r -d:nimDebugDlOpen examples\original\basic.nim
```

**Comment by greenfork** [comment]
@zetashift could you update the bindings from latest master and try again. It should just work. Only `raylib.dll` should be present in the same directory

**Comment by zetashift** [comment]
Running the `shapes/bouncing_ball` and the `shapes/draw_ring` example gives me on the latest master currently this error:
 ```
C:/Users/Rishi/.choosenim/toolchains/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/6.3.0/../../../../x86_64-w64-mingw32/bin/ld.exe: cannot find -lraylib
collect2.exe: error: ld returned 1 exit status
Error: execution of an external program failed: 'gcc.exe   -o P:\dev\gamedev\nimraylib_now\examples\shapes\bouncing_ball.exe  C:\Users\Rishi\nimcache\bouncing_ball_d\stdlib_io.nim.c.o C:\Users\Rishi\nimcache\bouncing_ball_d\stdlib_system.nim.c.o C:\Users\Rishi\nimcache\bouncing_ball_d\stdlib_dynlib.nim.c.o C:\Users\Rishi\nimcache\bouncing_ball_d\stdlib_winlean.nim.c.o C:\Users\Rishi\nimcache\bouncing_ball_d\stdlib_times.nim.c.o C:\Users\Rishi\nimcache\bouncing_ball_d\stdlib_os.nim.c.o C:\Users\Rishi\nimcache\bouncing_ball_d\@m..@s..@ssrc@snimraylib_now@sraylib.nim.c.o C:\Users\Rishi\nimcache\bouncing_ball_d\@mbouncing_ball.nim.c.o  -lraylib   '
```

I think I have to pass in the raylib headers somewhere then?

**Comment by greenfork** [comment]
Yes, here is the [line](https://github.com/greenfork/nimraylib_now/blob/master/src/nimraylib_now/raylib.nim#L8):
```nim
{.passL:"-lraylib".}
```

Possible problem is that mingw (gcc.exe) doesn't see the raylib.dll file. Possible solution: -L flag
```shell
nim r --passL:"-L<path/to/directory>" -d:nimDebugDlOpen examples\original\basic.nim
```

Another solution is to change `{.passL:"-lraylib".}` to something else, but I'm not sure what exactly

Searches on the topic, nothing seems to have the exact answer but for reference:
- https://stackoverflow.com/questions/1985764/how-to-add-3rd-party-libraries-to-mingw
- https://stackoverflow.com/questions/43311621/c-correct-way-to-statically-dynamically-link-with-mingw-w64
- https://docs.microsoft.com/en-us/troubleshoot/windows-client/deployment/dynamic-link-library

Mingw website is currently down for some reason, would be good to browse their wiki once it's up e.g. this one http://www.mingw.org/wiki/LibraryPathHOWTO

**Comment by greenfork** [comment]
According to these answers:
- https://stackoverflow.com/questions/185326/whats-a-dll-a-file
- https://stackoverflow.com/q/13998575/8598954
And wiki:
- https://en.wikipedia.org/wiki/Dynamic-link_library#Import_libraries

Seems that libraylibdll.a is also needed for linking

Raylib ships with 3 binaries:
```
raylib-3.5.0_win64_mingw-w64/lib/libraylib.a
raylib-3.5.0_win64_mingw-w64/lib/libraylibdll.a
raylib-3.5.0_win64_mingw-w64/lib/raylib.dll
```

and I assume that you need both libraylibdll.a and raylib.dll present.

Alternatively we might want to pass it like this "--passL:libraylibdll.a" as @enthus1ast did it here https://github.com/enthus1ast/raylib/blob/master/src/draylib.nim#L4

**Comment by greenfork** [comment]
I remove this passL pragma from master so you have easier time tinkering with it. For me it now works like this:
```
nim r --passL:-lraylib examples/original/crown.nim
```

My current understanding is that one of these options should work for Windows if you have libray.dll and libraydll.a files in the same directory as exe:
```
nim r --passL:"-lraylib" examples\original\crown.nim
nim r --passL:"-L/path/to/exe/folder -lraylib" examples\original\crown.nim
nim r --passL:"libraylibdll.a" examples\original\crown.nim
nim r --passL:"-L/path/to/exe/folder libraylibdll.a" examples\original\crown.nim
```

Please let me know if one of these commands work or you find anything else. I think I will be able to automate path finding after we find the correct command and flags

**Comment by zetashift** [comment]
`nim r --passL:"libraylibdll.a" examples\original\crown.nim` works! As long as `libraylibdll.a` is in the same folder. `nim r --passL:"-L/path/to/exe/folder libraylibdll.a" examples\original\crown.nim` works too! 

```
nim r --passL:"-lraylib" examples\original\crown.nim
nim r --passL:"-L/path/to/exe/folder -lraylib" examples\original\crown.nim
nim r --passL:"libraylibdll.a" examples\original\crown.nim
nim r --passL:"-L/path/to/exe/folder libraylibdll.a" examples\original\crown.nim
```
1. The first one errored out with: `ld.exe: cannot find -lraylib
collect2.exe: error: ld returned 1 exit status`
2. Errors out with the same as above.
3. Works whoop whoop!
4. Works too!

The raygui example I tried worked too.

**Comment by zetashift** [comment]
`nimble testExamples` still gives me the error:
```
PS P:\dev\gamedev\nimraylib_now> nimble testExamples
  Executing task testExamples in P:\dev\gamedev\nimraylib_now\nimraylib_now.nimble
no tests were found for pattern: 'tests\examples\**\*.nim'
```

**Comment by zetashift** [comment]
Running a single example using `testament r` I get this monstrosity:
https://play.nim-lang.org/#ix=2LLJ

**Comment by greenfork** [comment]
Thank you very much for the feedback, I'm glad we finally found the right flag!

Current master fixes problem with passL flags like this in [raylib.nim](https://github.com/greenfork/nimraylib_now/blob/master/src/nimraylib_now/raylib.nim):
```nim
when defined(windows):
  when defined(vcc):
    # Should it be `link` instead of passL?
    {.passL:"raylibdll.lib".}
  else:
    {.passL:"libraylibdll.a".}
else:
  {.passL:"-lraylib".}
```

And "no tests were found for pattern" problem by replacing "**" glob to just single "*" asterisk since double asterisk is not present in power shell, so the testament command is fixed to be like this:
```nim
task testExamples, "checks that all examples are correctly compiled":
  let pattern = "'tests"/"examples"/"*"/"*.nim'"
  exec "testament pattern " & pattern
```

In short, everything should work in regards to running and compiling any code if dll files are in the same directory as exe.

I don't know how to solve the problem with testament output though.

Let me know if the latest master works for you and I will release a new version and update readme accordingly.

**Comment by zetashift** [comment]
Samples I have tried, like crown/draw_ring/bouncing_ball work great on latest master by simply running:
`nim c -r foo.nim` ofcourse having the .dll and .a file in the same folder as resulting .exe.

Thank you!!

**Comment by greenfork** [comment]
Great, thanks! A new version will be released soon

**Comment by zetashift** [comment]
I can finally try this on my Linux laptop :P. 
I'll close this, the examples might be a whole other issue.

**Comment by greenfork** [comment]
Mission completed indeed. Thank you for your time!
